File: asn1runtime/src/org/asn1gen/runtime/AsnOctetString.scala
package org.asn1gen.runtime

class AsnOctetString(val value: List[Byte]) extends AsnType {
  override def _desc: meta.AsnOctetString = meta.AsnOctetString

  def copy(value: List[Byte] = this.value) = new AsnOctetString(value)

  override def equals(that: Any): Boolean = that match {
    case that: AsnOctetString => this.value == that.value
    case _ => false
  }

  override def hashCode(): Int = this.value.hashCode

  def value(f: (List[Byte] => List[Byte])): AsnOctetString = this.copy(value = f(this.value))

  def unapply(): Option[(List[Byte])] = Some(value)
  
  def string: String = new String(this.value.toArray)
}

object AsnOctetString extends AsnOctetString(Nil) {
  def apply(value: List[Byte]): AsnOctetString = new AsnOctetString(value)
  
  def apply(value: String): AsnOctetString = AsnOctetString(value.getBytes.toList)
}

File: asn1jrt/src/main/java/org/asn1gen/runtime/java/AsnOctetString.java
package org.asn1gen.runtime.java;

import java.nio.charset.Charset;

public class AsnOctetString implements AsnType {
  public static final AsnOctetString EMPTY = new AsnOctetString(new byte[0]);
  
  public final byte[] value;
  
  public AsnOctetString(final byte[] value) {
    this.value = value;
  }
  
  public String value(final Charset charset) {
    return new String(this.value, charset);
  }
}
